Handle ITupleBinaryOperation in CFG

mavasani opened the pull request.

jcouv commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1458,6 +1458,14 @@ public override IOperation VisitBinaryOperator(IBinaryOperation operation, int?
+            IOperation rightOperand = Visit(operation.RightOperand);
+            return new TupleBinaryOperatorExpression(operation.OperatorKind, _evalStack.Pop(), rightOperand,
+                semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, IsImplicit(operation));
+        }
```
The ordering for `(a, b) == (c, d)` is `tempA = a; tempB = b; tempC = c; tempD = d; tempA == tempC && tempB == tempD`.
That last expression has a mini control-flow. This can skip some conversions and also some user-defined comparison operators. 
Do we expect to represent that in the CFG?

mavasani replied:
Based on our discussion, filed https://github.com/dotnet/roslyn/issues/26969 as a follow-up issue.

jaredpar replied:
Thanks for filing the issue. Definitely feels like something the CFG needs to express. There is no way we can expect consumers to be able to correctly infer the control flow + side effects here. Too complex. 

---
In reply to: [189291960](https://github.com/dotnet/roslyn/pull/26919#discussion_r189291960) [](ancestors = 189291960)

mavasani commented:
@dotnet-bot retest windows_debug_spanish_unit32_prtest  please

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1458,6 +1458,14 @@ public override IOperation VisitBinaryOperator(IBinaryOperation operation, int?
 
+        public override IOperation VisitTupleBinaryOperator(ITupleBinaryOperation operation, int? captureIdForResult)
+        {
+            _evalStack.Push(Visit(operation.LeftOperand));
```
>_evalStack.Push(Visit(operation.LeftOperand)); [](start = 12, length = 46)

If operands are tuple expressions, should we preserve those as we do for deconstruction, for example?

AlekseyTs replied:
If we do that, we probably should do that for the other side as well, but it looks like we are not doing this for deconstruction. Probably should do as well.

---
In reply to: [189724034](https://github.com/dotnet/roslyn/pull/26919#discussion_r189724034) [](ancestors = 189724034)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleBinaryOperatorExpression.cs:
```diff
@@ -195,5 +195,321 @@ void M1((int, int) t1, long l)
+            var expectedDiagnostics = DiagnosticDescription.None;
+
+            VerifyFlowGraphAndDiagnosticsForTest<BlockSyntax>(source, expectedFlowGraph, expectedDiagnostics);
+        }
```
> } [](start = 7, length = 2)

Please add tests when left and right are tuple operations and add some variations with control flow (given current implementation, no need to test more than one control flow construct on each side)

AlekseyTs commented:
Done with review pass (iteration 1)

mavasani commented:
@AlekseyTs addressed feedback

---
In reply to: [390796883](https://github.com/dotnet/roslyn/pull/26919#issuecomment-390796883) [](ancestors = 390796883)

mavasani commented:
Ping @jcouv @333fred @dotnet/roslyn-compiler for second review

333fred commented:
@mavasani test failures look legitimate

mavasani commented:
@333fred Thanks. Seems like merge commit introduced duplicate switch labels. Should be fixed now.
